snitun/multiplexer/queue.py

```diff
@@ -0,0 +1,181 @@
+"""Multiplexer message handling."""
+
+from __future__ import annotations
+
+import asyncio
+from collections import OrderedDict, defaultdict, deque
+import contextlib
+from dataclasses import dataclass, field
+
+from .message import MultiplexerChannelId, MultiplexerMessage
+
+
+@dataclass(slots=True)
+class _ChannelQueue:
+    """Channel queue."""
+
+    total_bytes: int = 0
+    queue: deque[MultiplexerMessage | None] = field(default_factory=deque)
+    putters: deque[asyncio.Future[None]] = field(default_factory=deque)
+
+
+class MultiplexerSingleChannelQueue(asyncio.Queue[MultiplexerMessage | None]):
+    """Multiplexer single channel queue.
+
+    qsize is the size of the queue in bytes instead of the number of items.
+    """
+
+    _total_bytes: int = 0
+
+    def _put(self, message: MultiplexerMessage | None) -> None:
+        """Put a message in the queue."""
+        size = 0 if message is None else len(message.data)
+        self._total_bytes += size
+        super()._put(message)
+
+    def _get(self) -> MultiplexerMessage | None:
+        """Get a message from the queue."""
+        message = super()._get()
+        size = 0 if message is None else len(message.data)
+        self._total_bytes -= size
+        return message
+
+    def qsize(self) -> int:
+        """Size of the queue in bytes."""
+        return self._total_bytes
+
+
+class MultiplexerMultiChannelQueue:
+    """Multiplexer queue."""
+
+    def __init__(self, channel_size_limit: int) -> None:
+        """Initialize Multiplexer Queue.
+
+        Args:
+            channel_size_limit (int): The maximum size of a channel
+            data queue in bytes.
+
+        """
+        self._channel_size_limit = channel_size_limit
+        self._channels: defaultdict[MultiplexerChannelId, _ChannelQueue] = defaultdict(
+            _ChannelQueue,
+        )
+        # order controls which channel_id to get next
+        self._order: OrderedDict[MultiplexerChannelId, None] = OrderedDict()
```

Comment:
Normal dict can be used since we didn't actually need the ability to reorder as we also round robin from the channels in get

Reply:
we need it for https://github.com/NabuCasa/snitun/pull/352/files#diff-e0707d490efd6dd23ad4451a60e2baf92d5fcdae364b9ad2bc6ae0ea44ad5561R155

Reply:
https://github.com/NabuCasa/snitun/pull/359